docs/components/authentication.md

```diff
@@ -0,0 +1,185 @@
+# Authentication
+
+The Authentication component allows you to to implement authentication methods which can simply update the request
+with authentication detail (for example by adding an `Authorization` header).
+This is useful when you have to send multiple requests to the same endpoint. Using an authentication implementation,
+these details can be separated from the actual requests.
+
+
+## Installation
+
+```
+$ composer require php-http/message
+```
+
+
+## Authentication methods
+
+General usage looks like the following:
+
+``` php
+$authentication = new AuthenticationMethod();
+
+/** @var Psr\Http\Message\RequestInterface */
+$authentication->authenticate($request);
+```
+
+
+### Basic Auth
+
+This authentication method accepts two parameters: username and password. Getters/Setters are provided by
+`Http\Message\Authentication\UserPasswordPair` trait.
+
+``` php
+use Http\Message\Authentication\BasicAuth;
+
+$authentication = new BasicAuth('username', 'password');
+
+// These details can be set later as well
+// There are also getters with the appropriate method names
+$authentication->setUsername('username');
+$authentication->setPassword('password');
```

Comment:
Also, shouldn't we keep the mutability of the Chain method? It's some sort of builder class after all.